### ISSUE TITLE
Keyboard and Battery for 840 G4 without Touchscreen

### ISSUE THREAD

**Issue by DomiDomian** [XXCoreRangerX/Hackintosh-HP-EliteBook-840-G4#1]
Hi!

First thank you very much for sharing your work. I was trying your efi, but i have problems with keyboard. it is not recognized. Also battery symbol in istat menus always shows 20 hours of remaining battery time. I tried with and without voodool2c

Maybe you can help me

**Comment by XXCoreRangerX** [comment]
Hey, I've modified in the EFI to use `VoodooInput` from **acidanthera** as the main one - that should fix the keyboard for non-touch laptop model. Download the latest release and let me know if it works for you. Make sure `VoodooInput`, `VoodooPS2Controller` and `VoodooPS2Keyboard` are enabled in the config.

For fixing the battery, you might try downgrading to **1.29** BIOS, but I don't guarantee it's going to work. Battery management **only works properly if CFG Lock is disabled**, which is currently **not possible** on most HP laptops, as it's not a feature listed in the BIOS, neither in the BIOS bin file.

**Comment by DomiDomian** [comment]
Thanks for your response, 
yesterday i was trying around a little bit and used smbus, rmi and voodoops2 from my repo. -this worked for me and now keyboard is functional. i also tried your voodoops2mouse and yes the trackpad worked, this was the very first time. -But it breaked the multigesture of the touchpad but it had a very good feature i was look for long. now you could double tab a window and moving it by touchpad without of clicking the physical buttons of the touchpad. would be very great to implemet this gesture (trackpad is not important for me)

The thing with the battery:
i tried downgrading to 1.29 but the same thing on istat menus, it always shows 20 hours of remaining power. so far no problem because in most cases those calculations were not right. now i give it a try and only use remaining power in percentages. 

Power stepping seems to work great

greetings from hungary

**Comment by XXCoreRangerX** [comment]
Hey! Good to hear it's working properly for you now.

Drag lock is a feature that VoodooRMI doesn't have. You can enable it under macOS in System Preferences → Accessibility → Pointer Control → Mouse & Trackpad → Trackpad Options for a similar effect.

You can try making a custom hot patch for the battery, but I don't think it's gonna fix problems with stats menu.

**Comment by DomiDomian** [comment]
ok, that was an easy way, but for me draglock is an important feature, was wondering the whole time why they should drop it ;) Fine! You also only have battery percentage and ne estimated time???? for me even in windows 10 its the same, no more time is shown, only for charging that also works on mac os

**Comment by XXCoreRangerX** [comment]
Yeah, estimated time doesn't work for me too. Adding `if osi darwin` to **SSDT-BATT** may fix it on Windows, but I don't know how to fix it on macOS yet.
